cmd/docsync/main.go

```diff
@@ -0,0 +1,169 @@
+// Command docsync uploads a local directory to a specified
+// S3 bucket
+//
+// Usage:
+//
+//     docsync bucket bucketPrefix localDir
+//
+// where bucket is the name of your S3 bucket, bucketPrefix is the
+// S3 prefix to be applied to all uploaded files, and localDir is a
+// directory containing compiled docs (from docgenerate) or other
+// files to upload.
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"mime"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func main() {
+	bucket := os.Args[1]
+	bucketPrefix := os.Args[2]
```

Comment:
I think it might be a good footgun prevention device to ensure the presence of a trailing slash here. I don't think it's ever valid for there not to be a trailing slash.

The tool I like for these situations is `strings.TrimPrefix`, e.g.:

    s = strings.TrimPrefix(s, "/") + "/"